packages/app/src/components/Layout/ScatterLayout/ScatterAxis.js

```diff
@@ -1,60 +1,179 @@
-import React from 'react'
+import React, { useRef, useState } from 'react'
 import PropTypes from 'prop-types'
 import { connect } from 'react-redux'
+import { Popover, FlyoutMenu, Tooltip, MenuItem, MenuDivider } from '@dhis2/ui'
+import i18n from '@dhis2/d2-i18n'
+import { AXIS_ID_COLUMNS, DIMENSION_ID_DATA } from '@dhis2/analytics'
 
 import Chip from '../Chip'
 import { sGetUi, sGetUiLayout } from '../../../reducers/ui'
 import { acSetUiActiveModalDialog } from '../../../actions/ui'
 import styles from './styles/ScatterAxis.style'
-import { AXIS_ID_COLUMNS, DIMENSION_ID_DATA } from '@dhis2/analytics'
+import IconButton from '../../IconButton/IconButton'
+import MoreHorizontalIcon from '../../../assets/MoreHorizontalIcon'
+import {
+    ITEM_ATTRIBUTE_HORIZONTAL,
+    ITEM_ATTRIBUTE_VERTICAL,
+} from '../../../modules/ui'
+import {
+    acSetUiItemAttributes,
+    acRemoveUiItemAttributes,
+} from '../../../actions/ui'
+import { sGetUiItemsByAttribute } from '../../../reducers/ui'
 
-class Axis extends React.Component {
-    render() {
-        const {
-            label,
-            style,
-            getOpenHandler,
-            items,
-            itemAttribute,
-        } = this.props
+const Axis = ({
+    label,
+    style,
+    getOpenHandler,
+    items,
+    itemAttribute,
+    getItemsByAttribute,
+    setItemAttributes,
+    removeItemAttributes,
+}) => {
+    const buttonRef = useRef()
+    const [dialogIsOpen, setDialogIsOpen] = useState(false)
 
-        return (
-            <div
-                id={label}
-                data-test={`${label}-axis`}
-                style={{ ...styles.axisContainer, ...style }}
-                onDragOver={this.onDragOver}
-            >
-                <div style={styles.label}>{label}</div>
-                <Chip
-                    axisId={AXIS_ID_COLUMNS}
-                    axisName={label}
-                    onClick={getOpenHandler(itemAttribute)}
-                    dimensionId={DIMENSION_ID_DATA}
-                    items={items}
-                    isLocked={true}
-                />
-            </div>
+    const toggleMenu = () => setDialogIsOpen(!dialogIsOpen)
+
+    const destination =
+        itemAttribute === ITEM_ATTRIBUTE_VERTICAL
+            ? ITEM_ATTRIBUTE_HORIZONTAL
+            : ITEM_ATTRIBUTE_VERTICAL
+
+    const onSwap = () => {
+        toggleMenu()
+        const verticalItems = getItemsByAttribute(ITEM_ATTRIBUTE_VERTICAL)
+        const horizontalItems = getItemsByAttribute(ITEM_ATTRIBUTE_HORIZONTAL)
+        removeItemAttributes(
+            DIMENSION_ID_DATA,
+            verticalItems,
+            ITEM_ATTRIBUTE_VERTICAL
+        )
+        removeItemAttributes(
+            DIMENSION_ID_DATA,
+            horizontalItems,
+            ITEM_ATTRIBUTE_HORIZONTAL
+        )
+        setItemAttributes(
+            DIMENSION_ID_DATA,
+            verticalItems,
+            ITEM_ATTRIBUTE_HORIZONTAL
+        )
+        setItemAttributes(
```

Comment:
Is this causing multiple re-renders in components where these item attributes are used?

Reply:
Good question, I'm not exactly sure how this works. I only "see" one update, but maybe there are actually 4 re-renders bundled as one?
Maybe @janhenrikoverland or @amcgee could weigh in? 
The option I guess is to create a new action e.g. `acSwapItemAttributes` which handles this in one go?